[WIP] Banker Enhancement

mckinlee opened the pull request:
This is not a port. This makes the beggar NPC act like the banker from MM. All excess rupees from checks are automatically added to the players bank balance. Visit the banker to make a withdrawal in Castle Town as child or Kakariko as adult.

Currently only working for Windows.

Not all features are available yet, but this is a functioning prototype. Many kinks need to be worked.

-You have the ability to make a deposit now

Feel free to give it a test should you have Windows.

https://imgur.com/a/ftlQEWl

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253689516.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253699690.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253700827.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253701075.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253703822.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253705094.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1253722477.zip)
<!--- section:artifacts:end -->

mckinlee committed:
WIP Banker Enhancement

This enahncement introduces a banker similiar to the one in MM.
The enhancement is functional, but missing some features I plan to add soon.

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-banker

mckinlee committed:
Added deposit feature

mckinlee committed:
Merge branch 'develop-banker' of https://github.com/mckinlee/Shipwright into develop-banker

PeeMoldy commented:
> This is not a port. This makes the beggar NPC act like the banker from MM. All excess rupees from checks are automatically added to the players bank balance. Visit the banker to make a withdrawal in Castle Town as child or Kakariko as adult.
> 
> Currently only working for Windows.
> 
> Not all features are available yet, but this is a functioning prototype. Many kinks need to be worked.
> 
> -You have the ability to make a deposit now
> 
> Feel free to give it a test should you have Windows.
> 
> https://imgur.com/a/ftlQEWl
> 
> ### Build Artifacts
> * [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1102773093.zip)
> * [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1102773098.zip)

That is really cool. Do you think you can implement other MM features into SoH? Like as requested here?

https://github.com/HarbourMasters/Shipwright/issues/3475

mckinlee commented:
> > This is not a port. This makes the beggar NPC act like the banker from MM. All excess rupees from checks are automatically added to the players bank balance. Visit the banker to make a withdrawal in Castle Town as child or Kakariko as adult.
> > 
> > Currently only working for Windows.
> > 
> > Not all features are available yet, but this is a functioning prototype. Many kinks need to be worked.
> > 
> > -You have the ability to make a deposit now
> > 
> > Feel free to give it a test should you have Windows.
> > 
> > https://imgur.com/a/ftlQEWl
> > 
> > ### Build Artifacts
> > * [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1102773093.zip)
> > * [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1102773098.zip)
> 
> That is really cool. Do you think you can implement other MM features into SoH? Like as requested here?
> 
> https://github.com/HarbourMasters/Shipwright/issues/3475

Glad ya like it, but probably not if I'm being honest. They just don't interest me, but that doesn't mean someone else won't do it down the line!

mckinlee committed:
Dialogue Update and Restoring Deleted Code

-Restored the original accumulator functionality with a czar
-Updated the dialogue now that depositing is a option

mckinlee commented on soh/src/code/z_parameter.c:
```diff
@@ -6522,14 +6522,14 @@ void Interface_Update(PlayState* play) {
+            if (gSaveContext.rupeeAccumulator > 0) {
+                gSaveContext.playerBalance += gSaveContext.rupeeAccumulator;
                 gSaveContext.rupeeAccumulator = 0;
             }
```
Restored original rupee handling code and made new excess rupee storage feature optional via 'gBanker' setting with the latest commit.

mckinlee committed:
A few updates
-Cleaned up Banker.c
-Added some new save context that I'll need later, but haven't used yet
-Added a lot for dialogue that isn't used yet.
-Removed reference to Pirate's Charm in codebase because I'll be implementing it as a item later on.

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-banker

mckinlee committed:
Clean up
-Banker.c was a mess so I reworked it.
-I have no idea when I'll get around to finishing this project. I'm
having a hard time figuring out how to do the rewards system go figure.

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-banker

mckinlee committed:
Missing Includes
-Added stdlib and functions.h headers. I think this is why linux build
won't compile. Since I don't have a linux environment to test on I am
doing this commit to test if github actions will compile for linux now.

mckinlee committed:
Added missing declaration
-Declared BankerMain in z_en_hy. Maybe it'll compile in linux now...

mckinlee committed:
Reward Logic
You cannot actually receive the rewards just yet, however, the textboxes
will now display the correct text for the rewards you would receive.

mckinlee committed:
Added condition
Adds a conditional check at the beginning of the BankerMain function to
return early if the msgMode is MSGMODE_NONE. Theres no need for it to be
running otherwise.

mckinlee committed:
Revamp Textbox Handling
-I've completely redone the entire textbox handling for the banker. This
makes the textboxes way easier to manage and makes them more asthetically
please.
-I also added the players bank balance to the save editor because
it was a pain to test things otherwise.

mckinlee commented:
This still remains a WIP, but It has been revamped a lot since the initial commit and wanted to provide a quick update on my progress. Here are some of the notable changes:

Renamed Functions: Changed UpdateDigitOverlay to UpdateBankerOverlay, making it sound more specific to the banker stuff.

Looping for Efficiency: I've added a loop in UpdateBankerOverlay for drawing digits. Much cleaner than the first version where each digit was drawn separately.

Simpler Color Code: In the new UpdateBankerOverlay, setting the color is slicker - just one line with a ternary operator, instead of the if-else thing in the old DrawDigit.

Better Value Calculation: CalculateSelectedValue now has min and max limits compared to the older CalculateNewValue.

Upgraded Input Handling: ProcessInput in your new code does more stuff, like handling an extra text ID and smoother input cooldowns.

More Complex Banker Talk: HandleBankerInteraction has got way more going on with different text cases and stuff, compared to the simpler old version.

Overall Cleaner Code: Overall, the new code is neater, more organized, and just makes more sense.

mckinlee committed:
Added new custommesages and additional logic

mckinlee committed:
Added a extra check in banker.c and z_parameter.c

mckinlee committed:
Shot in the dark to make linux build

mckinlee committed:
All rewards implemented and linux/mac/wiiu/switch builds now work

mckinlee committed:
Merge branch 'develop' into develop-banker

mckinlee committed:
Merge branch 'develop' into develop-banker

mckinlee committed:
New dialogue
-Banker advises you of the fee now
-Banker speaks to you through the pirate's charm after a warp transfer

mckinlee committed:
Merge branch 'develop' into develop-banker

mckinlee committed:
Merge branch 'develop-banker' of https://github.com/mckinlee/Shipwright into develop-banker

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-banker

mckinlee committed:
Translations, Formatting, and Additional Dialogue
-All translations have been completely redone.
-All custom messages have been formatted correctly now.
-Added additional dialogue for when a warp transfer can't be done due to maxed bank balance.

mckinlee committed:
More Heart Piece Reward Logic
-When the player obtains the piece of heart they will now hold the piece
of heart over their head.
-Additional logic added to ensure the player retargets the banker actor

mckinlee committed:
Added func_80853148 to functions.h
-Windows build compiles fine with out this, but it's needed for the others.

mckinlee committed:
Rename func_80853148 to Player_StartTalking
-Will be doing a seperate PR for this as well.

mckinlee committed:
Remove call to Item_Give
-No need to manually give the item anymore

mckinlee committed:
Asthetic Changes
-Changed instances of Message_ContinueTextbox to Message_StartTextbox
when transitioning into a *AMMOUNT textbox. This way the growing
animation for textboxes is used instead.
-With the above change it made it where I could consistently set a
delayTimer for the overlay to be drawn making it appear around the same
time as the textbox text.
-I also added a fade in effect to the overlay. This makes it so the
overlay doesn't feel so jarring when it appears.

mckinlee committed:
Case Handling for AMOUNT textbox revamp
-What I had originally was just a complete mess. This is a lot more
organized and does fix a bug where the wrong textbox was showing when
trying to make a deposit with a maxed bank balance.

mckinlee committed:
CreateGetItemMessage

mckinlee committed:
Code Clean Up
-Reverted z_parameter.c to the latest commit on develop and added in logic
to add left over rupees to the players balance. It has parity with the
previous logic, but isn't as dumb. You could tell I had no idea what I
was doing a few months ago when I wrote it.
-Added reward checkboxes to the save editor.
-Added in a new custom message, but the logic has not been implemented yet.
-Made fee part of savecontext. I can't recall why, but I'll leave it.
Maybe I can use it later for randomizer.
-Several placeholders added with comments stating as such. The Pirate's
Charm item needs to be implemented.

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-banker

Kazamor commented:
Very nice work on this! It will be very useful in the randomizer. Some feedback based on some quick testing:

- The rupee fee doesn't apply to warp transfers. This might be intentional (or not finished yet?), but it seems a bit odd... It means you should never do a manual deposit because there is a fee for doing that while just letting your wallet overfill and warp transfer the excess is free (and more convenient). If you don't want the fee to apply to warp transfers, perhaps rebalance the rewards so the fee is removed before obtaining the warp transfers.

- Getting a message every single time I obtain rupees when my wallet is full gets very annoying. I don't think it's really necessary. Perhaps just the first time, just to inform the player that it happens. Or perhaps only when certain milestones are reached (1000, 2000, 3000, 4000, 5000).

- Not sure if this is intentional, but the banker goes back to the vanilla dialog ("sell me something with [C]") when you are wearing a mask. I thought it wasn't working at first because of this, and it took me a while to figure out it was because I was wearing the bunny hood.

- When trying to withdraw 0 rupees the banker responds by saying that you cannot ***deposit*** the air in your pockets.
- There are some very minor errors in the text. For example, sometimes you capitalize "Rupees" but sometimes you don't (the original game always does). When the banker asks, "What would you like to do?", one of the options is "Withdrawal Rupees" when it should be "Withdraw Rupees".

- I don't like the idea of getting a Piece of Heart for depositing 5000 rupees (currently you don't, but I see you want to add that eventually). The game already has the amount of heart pieces you need for 20 hearts, so adding more is kind of weird... The rewards should really only be unique rewards, such as the Pirate's Charm and fee removal. The bank is useful enough on its own, so we really don't need more incentive to use it.

I understand this is a work-in-progress, so you may already be aware of some/most of this. Just trying to help out.

When you do get the Pirate's Charm working, will it also be added to the randomizer item pool? That would be really nice.

mckinlee commented:
> Very nice work on this! It will be very useful in the randomizer. Some feedback based on some quick testing:
> 
> - The rupee fee doesn't apply to warp transfers. This might be intentional (or not finished yet?), but it seems a bit odd... It means you should never do a manual deposit because there is a fee for doing that while just letting your wallet overfill and warp transfer the excess is free (and more convenient). If you don't want the fee to apply to warp transfers, perhaps rebalance the rewards so the fee is removed before obtaining the warp transfers.
> 
> - Getting a message every single time I obtain rupees when my wallet is full gets very annoying. I don't think it's really necessary. Perhaps just the first time, just to inform the player that it happens. Or perhaps only when certain milestones are reached (1000, 2000, 3000, 4000, 5000).
> 
> - Not sure if this is intentional, but the banker goes back to the vanilla dialog ("sell me something with [C]") when you are wearing a mask. I thought it wasn't working at first because of this, and it took me a while to figure out it was because I was wearing the bunny hood.
> 
> - When trying to withdraw 0 rupees the banker responds by saying that you cannot ***deposit*** the air in your pockets.
> - There are some very minor errors in the text. For example, sometimes you capitalize "Rupees" but sometimes you don't (the original game always does). When the banker asks, "What would you like to do?", one of the options is "Withdrawal Rupees" when it should be "Withdraw Rupees".
> 
> - I don't like the idea of getting a Piece of Heart for depositing 5000 rupees (currently you don't, but I see you want to add that eventually). The game already has the amount of heart pieces you need for 20 hearts, so adding more is kind of weird... The rewards should really only be unique rewards, such as the Pirate's Charm and fee removal. The bank is useful enough on its own, so we really don't need more incentive to use it.
> 
> I understand this is a work-in-progress, so you may already be aware of some/most of this. Just trying to help out.
> 
> When you do get the Pirate's Charm working, will it also be added to the randomizer item pool? That would be really nice.

Thank you for your feedback! It's really appreciated. The fee is planned to also apply for warp transfers. I just haven't worked out the logic for it just yet. The mask thing was totally not intentional and was pointed out in discord also. I plan to fix that in the next commit. I can totally see it getting a little annoying with the banker chiming in on every rupee check. That will definitely need to be reworked. I'd like to get your thoughts on this. 

What if for the following reasons:

- The very first warp transfer.
-If there is a fee, to confirm if you want to proceed with the warp transfer. The reason I think this might be needed is because in rando I want the fee to be random, similar to shop prices. Makes the rewards harder to get especially since I plan to randomize the rewards as well. This is also the reason I included a heart piece, I want to randomize that reward in addition to the pirates charm. 
-The first time the bank is to full. This one could be reset if you make a withdrawal so the next time you max it out he lets you know just one more time, just not every single time.
-at a certain milestone, unsure what yet, I do want him to give a hint through the charm of the whereabouts of a progressive wallet. He will only say it once. Optional through settings of course. 

Anyway, dialogue definitely still needs more work. And I'm curious what you think the 5000 rupee reward could be considering randomizer? If not a heart piece, maybe just 5 rupees like he does in MM for the 1000 reward? 

Oh you mentioned that you currently don't receive the reward for the piece of heart. That makes me believe you are on a out dated build. In the latest build, you do! I emplore you to check that version out. It's been greatly improved upon. 

Again, thank you for your feedback!

Kazamor commented:
> If there is a fee, to confirm if you want to proceed with the warp transfer. The reason I think this might be needed is because in rando I want the fee to be random, similar to shop prices.

Are you saying you want the fee to change every single time a warp transfer is done or just be a random amount for the entire randomizer run? If the fee doesn't change throughout the run then you really only need to inform the player about it once because after that I can just understand that X amount of rupees will be taken out of  every transfer (or a transfer won't happen unless it's more than that amount).

One idea I just had: There are a few times throughout the game where a text box appears but doesn't prevent you from controlling Link. For example, when you enter the volcano without the Goron tunic equipped (if it's not disabled in SoH settings), a message appears on the screen saying it's too hot. You can still move around and do things while that message is on the screen, and it disappears after a few seconds. Would it be possible to trigger a text box like that for the banker as well? I think that would make the notifications a lot less annoying, if so.

> And I'm curious what you think the 5000 rupee reward could be considering randomizer? If not a heart piece, maybe just 5 rupees like he does in MM for the 1000 reward?

Well, the tycoon's wallet upgrade is actually not obtainable at all in the vanilla game and doesn't have an "original location" in the randomizer. That would be a fitting place to add that upgrade and make it obtainable outside of the randomizer as a unique reward. In the randomizer it would be shuffled into the item pool, although maybe add an option for it not to be (so it would always give the tycoon's wallet). 5000 rupees is a lot to obtain and some people may not want to potentially be required to do it, just like they may not want the 100 skulltula token reward to be shuffled for that reason.

After the final reward is obtained, you could add the repeating 5 rupee reward like Majora's Mask, just for consistency. If you really want consistency with Majora's Mask, you should increase the maximum capacity of the bank to 6000, but not allow deposits/transfers if you have 5000 or more rupees in the bank. Not a big deal, just a nice little touch you could do if you want.

> Oh you mentioned that you currently don't receive the reward for the piece of heart. That makes me believe you are on a out dated build. In the latest build, you do! I emplore you to check that version out. It's been greatly improved upon.

I didn't think I got a heart piece the first time, but I just tested it again and I did. I am on the newest build.

mckinlee commented:
> Are you saying you want the fee to change every single time a warp transfer is done or just be a random amount for the entire randomizer run? If the fee doesn't change throughout the run then you really only need to inform the player about it once because after that I can just understand that X amount of rupees will be taken out of every transfer (or a transfer won't happen unless it's more than that amount).
> 

Yeh, it will just be a random amount at the beginning and never change after that. At least that's my plan anyway.

> One idea I just had: There are a few times throughout the game where a text box appears but doesn't prevent you from controlling Link. For example, when you enter the volcano without the Goron tunic equipped (if it's not disabled in SoH settings), a message appears on the screen saying it's too hot. You can still move around and do things while that message is on the screen, and it disappears after a few seconds. Would it be possible to trigger a text box like that for the banker as well? I think that would make the notifications a lot less annoying, if so.
> 

I love that idea and it would be doable.


> Well, the tycoon's wallet upgrade is actually not obtainable at all in the vanilla game and doesn't have an "original location" in the randomizer. That would be a fitting place to add that upgrade and make it obtainable outside of the randomizer as a unique reward. In the randomizer it would be shuffled into the item pool, although maybe add an option for it not to be (so it would always give the tycoon's wallet). 5000 rupees is a lot to obtain and some people may not want to potentially be required to do it, just like they may not want the 100 skulltula token reward to be shuffled for that reason.
> 

That is a great idea and can be a setting like the 100 skulls.

> After the final reward is obtained, you could add the repeating 5 rupee reward like Majora's Mask, just for consistency. If you really want consistency with Majora's Mask, you should increase the maximum capacity of the bank to 6000, but not allow deposits/transfers if you have 5000 or more rupees in the bank. Not a big deal, just a nice little touch you could do if you want.
> 

I could.... but sounds like a waste of time just to add that logic. I'd say, maybe once everything else works as it should.

> I didn't think I got a heart piece the first time, but I just tested it again and I did. I am on the newest build.

Great! At least you get a little something something to make it somewhat worth testing out for now.

mckinlee committed:
Merge branch 'develop' into develop-banker

Extloga commented on soh/soh/z_message_OTR.cpp:
```diff
@@ -187,4 +187,142 @@ extern "C" void OTRMessage_Init()
+    CustomMessageManager::Instance->CreateMessage(
+        customMessageTableID, TEXT_BANKER_OPTIONS,
+        CustomMessage("What would you like to do?" "\x1C" "&%gDeposit Rupees&%gWithdrawal Rupees&%gNothing%w",
+                    "Was würdest du gern tun?" "\x1C" "&%gRubine einzahlen&%gRubine abheben&%gNichts%w",
```
Considering the majoritarian usage of the polite form in the added dialogue lines, I would suggest to standardize the German dialogue in a consistent way.

`Was würdest du gern tun?` &#8594; `Was würden Sie gern tun?`

Extloga commented on soh/soh/z_message_OTR.cpp:
```diff
@@ -187,4 +187,142 @@ extern "C" void OTRMessage_Init()
+    CustomMessageManager::Instance->CreateMessage(
+        customMessageTableID, TEXT_BANKER_WITHDRAWAL_CONFIRM,
+        CustomMessage("Here are your rupees. Your new&balance is %y{{playerBalance}}%w rupees. See you&later!",
+                      "Hier sind deine Rupien. Ihr neues&Guthaben beträgt %y{{playerBalance}}%w Rupien.&Bis später!",
```
Considering the majoritarian usage of the polite form in the added dialogue lines, I would suggest to standardize the German dialogue in a consistent way.

`Hier sind deine Rupien.` &#8594; `Hier sind Ihre Rupien.`

Extloga commented on soh/soh/z_message_OTR.cpp:
```diff
@@ -187,4 +187,142 @@ extern "C" void OTRMessage_Init()
+    CustomMessageManager::Instance->CreateMessage(
+        customMessageTableID, TEXT_BANKER_REWARD_WARP_TRANSFER_ITEM,
+        CustomMessage("To reward your thriftiness, I've got&a special trinket for you...The&%rPirate's Charm%w!",
+                    "Um deine Sparsamkeit zu belohnen,&habe ich ein besonderes&Schmuckstück für dich... den&%rPiratenzauber%w!",
```
Considering the majoritarian usage of the polite form in the added dialogue lines, I would suggest to standardize the German dialogue in a consistent way.

`Um deine Sparsamkeit zu belohnen` &#8594; `Um Ihre Sparsamkeit zu belohnen`

`Schmuckstück für dich` &#8594; `Schmuckstück für Sie`

Extloga commented on soh/soh/z_message_OTR.cpp:
```diff
@@ -187,4 +187,142 @@ extern "C" void OTRMessage_Init()
+        //ITEM_RUPEE_BLUE is a placeholder for ITEM_PIRATE_CHARM which has not been implemented yet.
+        customMessageTableID, (GetItemID)TEXT_BANKER_REWARD_WARP_TRANSFER_LORE_1, ITEM_RUPEE_BLUE,
+        CustomMessage("This charm, infused with pirate&magic, can warp your treasures&directly to the bank! No more&heavy wallets for you!",
+                    "Dieser mit Piratenmagie&durchdrungene Zauber kann Ihre&Schätze direkt an die Bank bringen!&Keine schweren Geldbörsen mehr!",
```
I suggest a different preposition for the German translation.

`an die Bank` &#8594; `in die Bank`

This alternative would be in accordance with the English text.

`an die Bank` &#8594; `zur Bank`

Extloga commented on soh/soh/z_message_OTR.cpp:
The last sentence is in need of improvement.

`Reduzieren Sie es ein wenig, oder?` &#8594; `Reduzieren Sie ihn ein wenig, ja?`

Extloga commented on soh/soh/z_message_OTR.cpp:
Considering the majoritarian usage of the polite form in the added dialogue lines, I would suggest to standardize the German dialogue in a consistent way.

`Du würdest Rubine verlieren!` &#8594; `Sie würden Rubine verlieren!`

Extloga commented:
There is no consistency regarding the name of the currency in the German dialogue lines. A standardization to "Rupie" and "Rupien" or "Rubin" and "Rubine" is needed. I would highly prefer the latter variant.

mckinlee commented:
> There is no consistency regarding the name of the currency in the German dialogue lines. A standardization to "Rupie" and "Rupien" or "Rubin" and "Rubine" is needed. I would highly prefer the latter variant.

Appreciate the feedback. Should I continue this project I will implement those changes to the dialog.

mckinlee commented:
I am closing this PR. I had a lot of fun making it, but I've just kinda lost interest. Someone is welcome to pick up from where i left off if they so choose though.
